File: src/edu/ncsu/feddgame/GameInstance.java
package edu.ncsu.feddgame;

import static org.lwjgl.glfw.GLFW.GLFW_RESIZABLE;
import static org.lwjgl.glfw.GLFW.GLFW_TRUE;
import static org.lwjgl.glfw.GLFW.GLFW_VISIBLE;
import static org.lwjgl.glfw.GLFW.glfwInit;
import static org.lwjgl.glfw.GLFW.glfwWindowHint;
import static org.lwjgl.opengl.GL11.GL_COLOR_BUFFER_BIT;
import static org.lwjgl.opengl.GL11.GL_TEXTURE_2D;
import static org.lwjgl.opengl.GL11.glClear;
import static org.lwjgl.opengl.GL11.glEnable;

import org.joml.Matrix4f;
import org.joml.Vector3f;
import org.lwjgl.opengl.GL;

import edu.ncsu.feddgame.level.ILevel;
import edu.ncsu.feddgame.level.TestLevel;
import edu.ncsu.feddgame.render.Camera;
import edu.ncsu.feddgame.render.FloatColor;
import edu.ncsu.feddgame.render.Font;
import edu.ncsu.feddgame.render.Shader;
import edu.ncsu.feddgame.render.Texture;

public class GameInstance {
	
	public static Window window;
	public static ObjectManager objectManager;
	boolean canRender;
	public ILevel level;
	public static Shader shader;
	
	private static State state;
	private volatile boolean gameState; // Boolean value to pause logic Thread when state != GAME
	
	// Game begins here
	public GameInstance() {
		try { 	// Run and quit on error
			setup();
			renderLoop();
		} finally {
			//glfwTerminate();
			//System.exit(1);
		}
	}
	
	private void setup() { 	// Setup all the window settings
		//Window.setCallbacks();
		
		objectManager = new ObjectManager();
		
		if (!glfwInit()) 	//Throw exception if glfw fails to initialize
			throw new IllegalStateException("Can not initialize GLFW");

		glfwWindowHint(GLFW_RESIZABLE, GLFW_TRUE); 	// Set window resizable and visible (set at defaults right now)
		glfwWindowHint(GLFW_VISIBLE, GLFW_TRUE);
		
		window = new Window(800, 800, "Laser Amazer", false);
		
		state = State.MAIN_MENU; // Set the game state
		
		level = new TestLevel();
	}
	
	private void renderLoop() { 	// Render Loop
		GL.createCapabilities();
		
		//TODO Should probably throw exception and exit here if window is null
		
		//TODO: Move this to it's own class, and have Fonts instantiated with location info
		// Set up main menu text
		Font menuTitle = new Font("Laser Amazer", new FloatColor(255, 25, 0));
		Font menuItem = new Font("Start Game", new FloatColor(25, 255, 0));
		Font startGame = new Font("Press Space to Start Game", new FloatColor(0, 25, 255));
		
		Camera camera = new Camera(window.getWidth(), window.getHeight());
		
		glEnable(GL_TEXTURE_2D);
		
		level.renderObjects(); 	//Add all objects to the scene from the level class
		objectManager.updateModels();
		window.addElements();
		
		shader = new Shader("shader");
		Texture tex = new Texture("bound.png");
		
		Matrix4f scale = new Matrix4f()
				.translate(new Vector3f(100, 0, 0))
				.scale(40);
		Matrix4f target = new Matrix4f();
		
		camera.setPosition(new Vector3f(-100, 0, 0));
		
		double frameCap = 1.0 / 60; // 60 FPS
		
		double frameTime = 0;
		@SuppressWarnings("unused")
		int frames = 0;
		
		double time = getTime();
		double unprocessed = 0;
		
		new Thread(() -> logicLoop()).start(); 	//Run the logic in a separate thread
		
		while (!window.shouldClose()) { 	// Poll window while window isn't about to close
			canRender = false;
			
			// Control frames per second
			{
				double timeNow = getTime();
				double elapsed = timeNow - time;
				unprocessed += elapsed;
				frameTime += elapsed;
				time = timeNow;
			}
			
			// Run all non-render related tasks
			while (unprocessed >= frameCap) {
				unprocessed -= frameCap;
				canRender = true;
				target = scale;
				
				window.update();
				
				if (frameTime >= 1.0) {
					frameTime = 0;
					// Display FPS counter in console
					//System.out.println("FPS: " + frames);
					frames = 0;
				}
			}
			
			// Render when scene changes
			if (canRender) {
				glClear(GL_COLOR_BUFFER_BIT);
				//glClearColor(1, 1, 1, 1);
				
				if (state.equals(State.CREDITS)) {
					gameState = false;
				} else if (state.equals(State.GAME)) {
					gameState = true;
					shader.bind();
					shader.setUniform("sampler", 0);
					shader.setUniform("projection", camera.getProjection().mul(target));
					tex.bind(0);
					objectManager.renderAll();
					level.renderLoop();
					window.renderElements();
				} else if (state.equals(State.MAIN_MENU)) {
					gameState = false;
					shader.unbind();
					menuTitle.renderString(menuTitle.getRenderString(), 16, -0.82f, 0.3f, 0.4f);
					menuItem.renderString("> Start Game", 16, -0.64f, 0.1f, 0.3f);
					menuItem.renderString("> How to Play", 16, -0.64f, 0.02f, 0.3f);
					menuItem.renderString("> Credits", 16, -0.64f, -0.06f, 0.3f);
					startGame.renderString("(Press Space.)", 16, -0.72f, -0.45f, 0.3f);
				}
				
				window.swapBuffers(); // Swap the render buffers
				frames++;
			}
			
		}
	}
	
	private void logicLoop() {	
		long timing = Math.round(1f / 60 * 1000f); 	//Get the number of milliseconds between frames based on 60 times a second
		
		while (!window.shouldClose()){
			if (!gameState) continue;
			
			double timeNow = getTime(); 	//Get time at the start of the loop
			
			level.logicLoop(); 	//Run the logic necessary for the level
			
			{
				long sleeptime = timing - (long)(getTime() - timeNow); 	//Sync the game loop to update at the refresh rate
				//System.out.println(sleeptime);
				try {
					Thread.sleep(sleeptime);
				} catch (InterruptedException e) {
					e.printStackTrace();
				}
			}	
				
		}
	}
	
	/**
	 * @return System time in seconds
	 */
	public double getTime() {
		return (double) System.nanoTime() / (double) 1000000000L;
	}
	
	public static void setState(State s) {
		state = s;
	}
	
	public static State getState() {
		return state;
	}
	
}

File: FEDDGame/src/edu/ncsu/feddgame/GameInstance.cs
using OpenTK.Graphics.OpenGL;
using System.Collections.Generic;
using System.Collections;
using System.Threading;
using OpenTK;
using System;
using System.Diagnostics;

public class GameInstance {
	
	public static Window window;
	public static ObjectManager objectManager;

	public static bool levelCompleteDialogue = true;
	public static bool showTimer = true;
	public static int samplingLevel = 4;
	public static int latestLevel = 5;
	public static int currentLevel = 2;
	
	private bool canRender;
	public static readonly bool demoMode = true;
	
	public static List<Scene> scenes = new List<Scene>() {
		new MainMenu(),
		new OptionsMenu(),
		new Level1(),
		new Level2(),
		new Level3(),
		new Level4(),
		new Level5(),
		new Level6(),
		new Level7(),
		new Level8(),
		new Level9(),
		new Level10()
	};
	
	public static int levNum = 0; // Start with 0
	private static float fadeN = 90f; // Amount of fade in degrees (0-90)
	private static bool hasLevel = false;
	public static Shader shader;
	private Stopwatch timer = new Stopwatch();

	private static State state;
	static volatile bool gameState; // Boolean value to pause logic Thread when state != GAME
	
	// Game begins here
	public GameInstance() {
		// Run and quit on error
		try { 	
			setup();
			renderLoop();
		} finally {
			//glfwTerminate();
			//System.exit(1);
		}
	}
	
	/**
	 * Setup all the window settings
	 */
	private void setup() {
		SaveGame.readData();
		new ModLoader();
		objectManager = new ObjectManager();
		
		// If glfw fails to initialize, throw exception 
		if (!glfwInit())
			throw new IllegalStateException("Cannot initialize GLFW");

		glfwWindowHint(GLFW_RESIZABLE, GLFW_TRUE); 	// Set window resizable and visible (set at defaults right now)
		glfwWindowHint(GLFW_VISIBLE, GLFW_TRUE);
		glfwWindowHint(GLFW_SAMPLES, 4);
		glfwWindowHint(GLFW_STENCIL_BITS, 4);
		
		window = new Window(800, 800, "Laser Amazer", false);
		
		if (demoMode)
			scenes.add(new DemoLevel());
		scenes.add(new GameComplete());
	}
	
	/**
	 * Game render loop
	 */
	private void renderLoop() {
		GL.createCapabilities();
		//TODO Should probably throw exception and exit here if window is null
		
		glEnable(GL_TEXTURE_2D);
		glEnable(GL_BLEND);
		glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
		
		FloatColor clearColor = new FloatColor(GameColor.DARK_GREY.getColor());
		glClearColor(clearColor.red(), clearColor.blue(), clearColor.green(), 1f);

		// Set up main menu text
		GameFont menuItem = new GameFont("Start Game", new FloatColor(GameColor.ORANGE.getColor()));
		GameFont startGame = new GameFont("Press Space to Start Game", new FloatColor(GameColor.YELLOW.getColor()));
		
		Matrix4d scale = new Matrix4d().translate(new Vector3f(100, 0, 0)).scale(40);
		Matrix4d target = new Matrix4d();
		
		Camera camera = new Camera(window.getWidth(), window.getHeight());
		camera.setPosition(new Vector3d(-100, 0, 0));
		
		shader = new Shader("shader");
		
		double frameCap = 1.0 / 60; // 60 FPS
		
		double frameTime = 0;
		double time = getTime();
		double unprocessed = 0;
		
		setState(State.GAME); // Set the game state
		
		//
			setLevel(0); // Set the starting level
		//
			
		new Thread(() => logicLoop()).Start(); // Run the logic in a separate thread
		
		glfwSetWindowSize(window.window, 1200, 800);
		window.centerWindow(); // Center window on screen
		// Poll window while window isn't about to close
		while (!window.shouldClose()) {
			canRender = false;
			
			// Control frames per second
			{
				double timeNow = getTime();
				double elapsed = timeNow - time;
				unprocessed += elapsed;
				frameTime += elapsed;
				time = timeNow;
			}
			
			// Run all non-render related tasks
			while (unprocessed >= frameCap) {
				unprocessed -= frameCap;
				canRender = true;
				target = scale;
				
				window.update();
				
				if (frameTime >= 1.0) 
					frameTime = 0;
			}
			
			// Render when scene changes
			if (canRender) {
				renderLevel();
				glClear(GL_COLOR_BUFFER_BIT);
				
				if (state.equals(State.GAME)) {
					gameState = true;
					shader.bind();
					shader.updateUniforms(camera, target);
					objectManager.renderAll();
					window.updateTime();
					window.renderElements();
					getCurrentLevel().renderLoop();
				} else if (state.equals(State.LEVEL_COMPLETE)) {
					getCurrentLevel().setActive(false); // Set level as inactive
					
					// If user has the level complete dialogue enabled
					if (levelCompleteDialogue) {
						gameState = false;
						
						shader.bind();
						shader.updateUniforms(camera, target);
						objectManager.renderAll();
						
						window.renderElements();
						
						// Add dark rectangle to make text more readable
						shader.unbind();
						glEnable(GL_BLEND);
						glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
						
						glColor4f(clearColor.red(), clearColor.green(), clearColor.blue(), .5f);
						glRectf(-10f, -10f, 10f, 10f);
						
						Scene latestLevel = getCurrentLevel();
						
						menuItem.renderString("Congratulations!",  Alignment.CENTER, 0.1f, 0.45f);
						menuItem.renderString("You've completed " + latestLevel.getName() + " in " + latestLevel.getElapsedSeconds() + " seconds.", Alignment.CENTER, 0.02f, 0.2f);
						startGame.renderString("(Press Space to continue.)", Alignment.CENTER, -0.45f, 0.3f);
					} else {
						setState(State.NEXT_LEVEL);
					}
				} else if (state.equals(State.NEXT_LEVEL)) {
					gameState = false;
					nextLevel();
					setState(State.GAME);
				}
				
				if (!state.equals(State.LEVEL_COMPLETE)) {
					// Fading
					shader.unbind();
					glEnable(GL_BLEND);
					glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
					
					if (fade > 0) {
						fade -= 2.5f;
						glColor4f(clearColor.red(), clearColor.blue(), clearColor.green(), (float) Math.sin(Math.toRadians(fade)));
						glRectf(-10f, -10f, 10f, 10f);
					}
				}

				window.swapBuffers(); // Swap the render buffers
			}
			
		}
		SaveGame.writeData();
	}
	
	private void logicLoop() {
		Thread.CurrentThread.Name = "Logic";
		int timing = (int)Math.Round(1f / 60 * 1000f); 	// Get the number of milliseconds between frames based on 60 times a second
		
		while (!window.shouldClose()) {
			if (!gameState) continue;
			
			double timeNow = getTime(); // Get time at the start of the loop
			
			// Make sure it's the active level
			if (getCurrentLevel().isActive())
				getCurrentLevel().logicLoop();
			
			{
				int sleeptime = timing - (int)(getTime() - timeNow); // Sync the game loop to update at the refresh rate
				//System.out.println(sleeptime);
				try {
					Thread.Sleep((int)sleeptime);
				} catch (Exception e) {
				}
			}	
				
		}
	}
	
	/**
	 * @return System time in seconds
	 */
	public double getTime() {
		return timer.ElapsedMilliseconds;
	}
	
	public static void setState(State s) {
		fade();
		state = s;
	}
	
	public static State getState() {
		return state;
	}
	
	private static void nextLevel() {
		getCurrentLevel().setActive(false); // No longer the active level
		
		// If all levels complete, reset to level 0
		levNum++;
		latestLevel = (levNum <= latestLevel) ? (latestLevel) : (levNum);
		if (levNum > 1)
			currentLevel = levNum;
		
		if (levNum > scenes.Count - 1) {
			levNum = 0;
			setLevel(scenes.Count - 1); // Set to GameComplete (last level)
		}
		
		getCurrentLevel().setActive(true); // Set new active level
		hasLevel = false;
	}
	
	public static Scene getCurrentLevel() {
		return scenes[levNum];
	}
	
	public static void setLevel(int levelNumber) {
		if (levelNumber > scenes.Count - 1) {
			levNum = scenes.Count - 1;
		} else {
			levNum = levelNumber;
		}
		
		if (demoMode) {
			if (levNum > 1 && levNum != scenes.Count - 2)
				currentLevel = levNum;
		} else {
			if (levNum > 1)
				currentLevel = levNum;
		}
		
		fade();
		latestLevel = (levNum <= latestLevel) ? (latestLevel) : (levNum);
		scenes[levNum].setActive(true); // Set active level
		hasLevel = false;
	}
	
	private void renderLevel() {
		if (levNum < scenes.Count && !hasLevel) {
			objectManager.clearAll();
			window.clearElements();
			window.addElements();
			getCurrentLevel().renderObjects(); // Add all objects to the scene from the level class
			objectManager.updateModels();
			hasLevel = true;
		} else if (!hasLevel) {
			Console.WriteLine("End of game");
		}
	}
	
	private static void fade() {
		fadeN = 90f;
	}
	
}
